### PR TITLE
adding test cases for download_data.py

### PR DESCRIPTION
**Pull request by ttimbers**
Please review the `download_data.py` and `test_download_data.py`. Since `dowload_data.py` doesn't involve code related to checklist requirements, so I only add two defensive test to make sure command line argument errors can be handled gracefully. Please feel free to comment in the code if you have any questions or suggestions!

### CODE REVIEW

File: scripts/download_data.py [hunk 1]
```diff
@@ -14,6 +15,19 @@
...
+    filename_from_url = os.path.basename(url)
+
+    # check if URL exists, if not raise an error
+    if request.status_code != 200:
+        raise ValueError('The URL provided does not exist.')
+    
+    # check if the URL points to a zip file, if not raise an error  
+    #if request.headers['content-type'] != 'application/zip':
+    if filename_from_url[-4:] != '.zip':
+        raise ValueError('The URL provided does not point to a zip file.')
```
Comment (ttimbers) [inline scripts/download_data.py:29]: These are already in the `read_zip` function, we do not need to duplicate them.

**Comment by ttimbers** [comment]
So from Openja et al (2024 preprint):

> _"Error and Exception Condition (EEC): This testing strategy employs the concept of exception and error handling [23] to derive the test cases and help detect or signal the errors (or potential errors) that can occur during the running of the ML software system. It is used to detect problems that may arise in the sequence of statements through the exception-related code or by executing a specific statement to trigger an exception, usually using the throw or raise statement. The steps in EEC include understanding limitations in the ML artifacts – data/features, model, or code across the ML work"ow and trying to make these aspects as robust as possible. Unfortunately, there is no one-size-fits-all solution for addressing ML challenges. Depending on the nature of the ML problem, various error and exception conditions may need to be considered. For example, in the context of computer vision tasks, it is imperative to ensure that the input data satis!es the necessary quality criteria, including validity. Furthermore, the features need to conform to the expectations of the model. Moreover, DL models often exhibit complexity and opacity, rendering the features utilized in the model difficult to interpret (explainable) due to their non-trivial nature."_

> _"Error and exception analysis is a critical process within ML systems, integral for enhancing the performance of ML models and ensuring their proper functioning in production environments. The primary objective of error and exception condition analysis in ML is to identify any errors that were inadequately handled or remained undetected during model deployment. This comprehensive process entails evaluating various ML artifacts, including data, models, features, and code, throughout the entirety of the ML work"ow and fortifying their robustness. "_

**Comment by ttimbers** [comment]
Improvements to make on current code, we only want to handle this error: https://github.com/FixML/breast_cancer_predictor_py/blob/2592bfe6b74c5057ed796a47214c655a8110b6eb/src/read_zip.py#L34

So we should change the type, or name it, and then modify this exception to work only for that error: https://github.com/FixML/breast_cancer_predictor_py/blob/2592bfe6b74c5057ed796a47214c655a8110b6eb/scripts/download_data.py#L19

Ways to name the error:

```
except ValueError as e:
 If e == “The directory provided….”:
    ….
```

But we think we should actually split this into two checks:

1. If this path exists, but its not a directory, we want to just throw an error that says a file exists with that name, it its not a directory and we need a new directory name. Throw a `NotADirectoryError`

2. If the path does not exist, we want to handle this by creating the directory. Throw a `FileNotFoundError` which we handle via the `except` that makes a directory.

**Comment by WeilinHan8** [comment]
Hi team, 
I've updated the pr regarding above discussion, please let me know if you have any questions!

Weilin

File: scripts/download_data.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import click
 import os
 import sys
+import requests
```
Comment (ttimbers) [inline scripts/download_data.py:8]: We don't need this - this is in `read_zip`
```suggestion
```

File: scripts/download_data.py [hunk 3]
```diff
@@ -14,11 +15,18 @@
 
 def main(url, write_to):
     """Downloads data zip data from the web to a local filepath and extracts it."""
+    
+    request = requests.get(url)
+    filename_from_url = os.path.basename(url)
+
```
Comment (ttimbers) [inline scripts/download_data.py:21]: This is also redundant to what is inside `read_zip`, so we don't need this.
```suggestion
```

File: src/read_zip.py [hunk 4]
```diff
@@ -29,10 +29,14 @@ def read_zip(url, directory):
...
-    # check if the directory exists, if not raise an error
+    # check if the directory path exists, if not raise an error
+    if not os.path.exists(directory):
+        raise FileNotFoundError('The directory provided does not exist.')
+    
+    # check if the dirctory path provided is a directory, if not raise an error
     if not os.path.isdir(directory):
-        raise ValueError('The directory provided does not exist.')
-
+        raise NotADirectoryError('The directory path provided is not a directory')
```
Comment (ttimbers) [inline src/read_zip.py:38]: Let's improve the error message a little.
```suggestion
        raise NotADirectoryError('The directory path provided is not a directory, it is an existing file path. Please provide a path to a new, or existing directory.')
```

File: tests/test_read_zip.py [hunk 5]
```diff
@@ -99,6 +99,11 @@ def test_read_zip_error_on_nonzip_url():
...
 # if the  directory path provided does not exist
-def test_read_zip_error_on_missing_dir():
-    with pytest.raises(ValueError, match='The directory provided does not exist.'):
+def test_read_zip_error_on_nonexistent_dir():
+    with pytest.raises(FileNotFoundError, match='The directory provided does not exist.'):
         read_zip(url_txt_csv_zip, 'tests/test_zip_data3')
+
+# if the  directory path provided does not exist
+def test_read_zip_error_on_missing_dir():
+    with pytest.raises(NotADirectoryError, match='The directory path provided is not a directory'):
```
Comment (ttimbers) [inline tests/test_read_zip.py:108]: Editing to match the improved error message:
```suggestion
    with pytest.raises(NotADirectoryError, match='The directory path provided is not a directory, it is an existing file path. Please provide a path to a new, or existing directory.'):
```

**Review by ttimbers — COMMENTED** [review]
Otherwise this looks good to me. But I have not yet checked that it runs in the docker container. We should do this. We should also set up CI for our tests to make this job easier for us. That can be a separate PR.

**Comment by ttimbers** [comment]
LGTM!

**Pull request merged**